### ISSUE TITLE
Grouping support in ComboBox

### ISSUE THREAD

**Issue by niels9001** [Microsoft/microsoft-ui-xaml#33]
# Proposal: Add grouping to ComboBox

## Summary
Grouping items in ComboBox is currently not supported. It was in WPF, and worked great. It's a gap we can bridge and to make it easier to port over WPF LOB apps to UWP.

![groupedcombobox](https://user-images.githubusercontent.com/9866362/30957493-b5719ff8-a43a-11e7-9c46-043f5a3404b8.jpg)

## Usage Examples

![groupedcombobox](https://user-images.githubusercontent.com/9866362/30957493-b5719ff8-a43a-11e7-9c46-043f5a3404b8.jpg)

**Comment by SavoySchuler** [comment]
This proposal compliments the one I submitted for adding [multi-select functionality to ComboBox](https://github.com/Microsoft/microsoft-ui-xaml/issues/263). Should grouping support be implemented, group select/de-select would be a great addition to multi-select functionality for ComboBox.

**Comment by SavoySchuler** [comment]
@niels9001, I have updated this proposal with a Rationale and Functional Requirements section.

**Comment by mdtauk** [comment]
There should be styles and properties to allow the templating and styling of the Combobox Group Headings

**Comment by SavoySchuler** [comment]
@mdtauk I added your comment as Functional Requirement #2!

**Comment by ChainReactive** [comment]
I've had this XAML present but commented out for six years now since uncommenting it causes the app to crash:

**Comment by SavoySchuler** [comment]
@ChainReactive Did you ever implement a workaround in that time? I'd be eager to know if there's any pros or cons from existing 3rd party solutions you'd like us to be aware of!

**Comment by ChainReactive** [comment]
Nope but users have complained about the lack of grouping.

**Comment by MikeHillberg** [comment]
Grouping in combo box is a standard part of the Ribbon UI as well.

**Comment by SavoySchuler** [comment]
@niels9001, @ChainReactive, and @mdtauk, thank you all for getting this feature started!

# It has been approved and I have opened up a spec for it [here](https://github.com/Microsoft/microsoft-ui-xaml-specs/blob/user/savoyschuler/comboboxgrouping/active/comboboxgrouping/GroupingSupportInComboBox.md). 

We would be eager to see you all involved in our spec writing process also where you can start telling us more specifics about how you would like this feature to be implemented. It may be several months before we are able to fully commit PM & Dev resources to this feature, but your early engagement will still help jumpstart this development. Please let me know if you have any questions. I have added our default spec template and will jump into contribute when I can!

Please note that [Multi-Select Functionality for ComboBox](https://github.com/Microsoft/microsoft-ui-xaml/issues/263) has _also_ been approved and these developments will inform and impact each other.

**Comment by SavoySchuler** [comment]
I'll be taking this again to pitch to refine the scope once more. @niels9001, @ChainReactive, and @mdtauk, were any of you _**needing**_ collapsible or multi-tiered groups for a published or in production app? Or are these two features properly labeled as a "could"?

**Comment by niels9001** [comment]
I personally would be fine with a simple implementation, that lets me group certain items. And a styleable groupheader.

Collapsing and multi-tiered groups would be a nice-to-have!

**Comment by SavoySchuler** [comment]
Re-pitched and marking "multi-level/hierarchical groups" as a requirement "won't" due to Flyout + TreeView solving same scenarios more capably (noting this capability could be due for a refresh via a separate proposal). 

I am ready to revisit spec writing, but will be blocked on a dev resource commitment until a validation partner is located. @Niels9001, @ChainReactive, @mdtauk, could any of you point me to an app in your umbrella that you would like to use this feature in as soon as its available?

**Comment by mdtauk** [comment]
@SavoySchuler Can't cite an app specifically, but the Fluent Web ComboBox supports it, so that would bring parity, but there is also the Ribbon, which has headered sections in drop downs, and a UWP Ribbon has long been on the "coming soon" ticket.

![image](https://user-images.githubusercontent.com/7389110/62089532-40c90b00-b261-11e9-9254-94c7f91653c3.png)

![image](https://user-images.githubusercontent.com/7389110/62089556-5b9b7f80-b261-11e9-82a0-bdc488f029ac.png)

![image](https://user-images.githubusercontent.com/7389110/62089611-900f3b80-b261-11e9-8fd3-1ccb7683de96.png)

![image](https://user-images.githubusercontent.com/7389110/62089664-c2209d80-b261-11e9-85c5-86eaaf8d0077.png)

**Comment by SavoySchuler** [comment]
Awesome. 

Two customers direct messaged me to let me know they are planning to use this feature in their apps. As soon as I've wound down NumberBox, and as per the advice from @mrlacey, I will complete the first draft of the spec prior to publishing it on the spec repo - that way there's no time lost guessing what's an oversight vs. what's in-development. Until then, I welcome API, behavioral components, visual components, and accessibility consideration conversations here!

Regarding timeline, note that  the preview of this feature will require the release of WinUI 3.0.

**Comment by mdtauk** [comment]
> Awesome.
> 
> Two customers direct messaged me to let me know they are planning to use this feature in their apps. As soon as I've wound down NumberBox, and as per the advice from @mrlacey, I will complete the first draft of the spec prior to publishing it on the spec repo - that way there's no time lost guessing what's an oversight vs. what's in-development. Until then, I welcome API, behavioral components, visual components, and accessibility consideration conversations here!
> 
> Regarding timeline, note that the preview of this feature will require the release of WinUI 3.0.

**ComboBoxGroup** - A XAML element with a header property containing ComboBox Items

**ComboBoxGroup.Header** - Item that can not be hovered or selected, but is a header for grouping items

```xaml
<ComboBox>
      <ComboBoxGroup Header="First Heading">
            <x:String>Blue</x:String>
            <x:String>Green</x:String>
            <x:String>Red</x:String>
            <x:String>Yellow</x:String>
      </ComboBoxGroup>
      <ComboBoxGroup Header="Second Heading">
            <x:String>Yes</x:String>
            <x:String>YASSS!</x:String>
      </ComboBoxGroup>
</ComboBox>
```

ComboBoxGroup should probably be designed to allow developers to derive from it, and create their own Groups with logic as they need.

Because the ComboBox uses a Flyout to display its contents, perhaps this Group element, should be a **FlyoutGroup** and maybe even a **MenuFlyoutGroup** which automatically handles Context Menu Radio Buttons, Context Menu Separators, etc.

**Comment by ChainReactive** [comment]
During a recent presentation to a stakeholder, the particular `ComboBox`'s lack of grouping resurfaced as a usability issue.  The grouping is essential to the sales pitch.

> Regarding timeline, note that the preview of this feature will require the release of WinUI 3.0.

@SavoySchuler That timeline is disappointing.  If it depends on WinUI 3.0 then why not include it in 3.0 itself rather than later?

I tried tweaking the `ComboBox` to make it more usable in the meantime.  Another complaint is the drop down is too short.  I set `MaxDropDownHeight` to a large number, but it has no effect!  My googling indicates this is [a lingering problem](https://social.msdn.microsoft.com/Forums/en-US/b32a9cec-accc-472b-8fcc-1eac77aa3280) since 2012.  Implementing grouping will be useless to me, unless the drop down can be taller.  Can you include a fix in the scope here, or must I open a separate issue?

**Comment by michael-hawker** [comment]
Wanted to point out this is also a WPF gap and a request we've had in the [Toolkit here](https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/1517).

As a workaround, in the meantime, you can custom style the existing ComboBox template, swap the ScrollViewer>ItemsPresenter part to a ListView w/ Grouping enabled, and then bind the ListView's `ItemsSource` and `SelectedItem` as TemplateBindings. However, the Popup menu sticks around, so code is needed to dismiss it, especially on selection. Surprising as GroupStyle is still on ComboBox, it's just ignored.

**Comment by yoshiask** [comment]
> Regarding timeline, note that the preview of this feature will require the release of WinUI 3.0.

This is rather unfortunate, as I'm currently working on a Ribbon control for the 7.0 release of the Toolkit which really needs this. As it stands, it looks like I'll be making my own custom control for this (for the Gallery).

**Comment by mdtauk** [comment]
I think it does require the source to be checked in before work can be done to it (flyouts for form controls currently can not match the compact density of the parent control for this reason)

Grouping in ComboBoxes and other Drop Downs is present in Fluent UI on the Web, so this would bring it to parity.
![image](https://user-images.githubusercontent.com/7389110/87245187-523c0080-c43b-11ea-9451-1d9ee2956697.png)

**Comment by MaddalaN** [comment]
@michael-hawker is this feature ready? if not, please let us know the status or plan to support this feature? Thanks in advance.

**Comment by TimKAtos** [comment]
Anything new here?

**Comment by michael-hawker** [comment]
Wish my past self had posted my style, can't seem to find it anywhere. Couldn't get it to work again. The ComboBox seemed to accept groups and worked displaying it once, but now is consistently crashing for me in the CollectionViewSource being set to the ItemsControl source in the generated bindings with an InvalidCastException. Not sure what happened.

Trying to use a TreeView in a Flyout now I guess, but that's more complicated to setup, and seems like I can't use the same grouped collection I have directly...?

**Comment by xperiandri** [comment]
@SavoySchuler any updates or progress?

**Comment by castorix** [comment]
> As a workaround, in the meantime, you can custom style the existing ComboBox template, swap the ScrollViewer>ItemsPresenter part to a ListView w/ Grouping enabled, and then bind the ListView's `ItemsSource` and `SelectedItem` as TemplateBindings. However, the Popup menu sticks around, so code is needed to dismiss it, especially on selection. Surprising as GroupStyle is still on ComboBox, it's just ignored.

Or another workaround is to add disabled items and change their color.
For example with  :

` Application.Current.Resources["ComboBoxItemForegroundDisabled"] = new SolidColorBrush(Microsoft.UI.Colors.LimeGreen);`

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/22345506/015dc3e9-4770-4745-bddb-cfcf2a1f276a)

**Comment by weitzhandler** [comment]
@SavoySchuler on [Mar 18 2019](https://github.com/microsoft/microsoft-ui-xaml/issues/33#issuecomment-474118908):

> [@niels9001](https://github.com/niels9001), [@ChainReactive](https://github.com/ChainReactive), and [@mdtauk](https://github.com/mdtauk), thank you all for getting this feature started!
> It has been approved and I have opened up a spec for it here

Hello five years later. Any plans?

**Comment by xperiandri** [comment]
@SavoySchuler and the spec is empty...
